server/app/services/docker/streaming_executor.rb

```diff
@@ -0,0 +1,70 @@
+module Docker
+  class StreamingExecutor
+
+    # @param [Container] container
+    # @param [#send,#close] ws
+    def initialize(container, ws)
+      @container = container
+      @ws = ws
+      @client = RpcClient.new(container.host_node.node_id)
+    end
+
+    # Starts exec session with tty
+    def start_tty
+      create_session
+      subscribe_to_session
+      register_tty_ws_events
+    end
+
+    # Starts normal exec session (without tty)
+    def start 
+      create_session
+      subscribe_to_session
+      register_run_ws_events
+    end
+
+    def create_session
+      @exec_session = @client.request('/containers/create_exec', @container.container_id)
+    end
+
+    def subscribe_to_session
+      @subscription = MongoPubsub.subscribe("container_exec:#{@exec_session['id']}") do |data|
+        if data.has_key?('exit')
+          @ws.send(JSON.dump({ exit: data['exit'] }))
+          @ws.close
```

Comment:
This could also use the close `code` (1000 vs something else) with a JSON payload for the exit event?